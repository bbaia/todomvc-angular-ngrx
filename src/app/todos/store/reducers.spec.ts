import * as fromTodos from '.';

describe('todos reducers', () => {
  beforeEach(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2020-01-01'));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should load todos', () => {
    // Arrange
    const loadAction = fromTodos.loadAction();
    const loadSuccessAction = fromTodos.loadSuccessAction({
      todos: [
        {
          id: 0.234567891,
          text: 'Learn Angular',
          creationDate: new Date('2020-02-01'),
          completed: true,
        },
        {
          id: 0.345678912,
          text: 'Learn Jest',
          creationDate: new Date('2020-03-01'),
          completed: false,
        },
      ],
    });

    // Act
    const state1 = fromTodos.reducer(fromTodos.initialState, loadAction);
    const state2 = fromTodos.reducer(state1, loadSuccessAction);

    // Assert
    expect(state1).toMatchSnapshot();
    expect(state2).toMatchSnapshot();
  });

  it('should add todos', () => {
    // Arrange
    jest.spyOn(Math, 'random').mockReturnValue(0.1);
    const firstAddAction = fromTodos.addAction('Learn Angular');
    jest.spyOn(Math, 'random').mockReturnValue(0.2);
    const secondAddAction = fromTodos.addAction('Learn NgRx');
    jest.spyOn(Math, 'random').mockRestore();

    // Act
    const state1 = fromTodos.reducer(fromTodos.initialState, firstAddAction);
    const state2 = fromTodos.reducer(state1, secondAddAction);

    // Assert
    expect(state2).toMatchSnapshot();
  });

  it('should mark todos as complete', () => {
    // Arrange
    const initialState: fromTodos.State = {
      data: {
        entities: {
          0.07: {
            id: 0.07,
            text: 'Learn Jest',
            creationDate: new Date(),
            completed: false,
          },
        },
        ids: [0.07],
      },
      loading: false,
    };
    const toggleAction = fromTodos.toggleAction({ id: 0.07 });

    // Act
    const state = fromTodos.reducer(initialState, toggleAction);

    // Assert
    expect(state).toMatchSnapshot();
  });

  it('should delete todos', () => {
    // Arrange
    const initialState: fromTodos.State = {
      data: {
        entities: {
          0.07: {
            id: 0.07,
            text: 'Learn Jest',
            creationDate: new Date(),
            completed: false,
          },
        },
        ids: [0.07],
      },
      loading: false,
    };
    const deleteAction = fromTodos.deleteAction({ id: 0.07 });

    // Act
    const state = fromTodos.reducer(initialState, deleteAction);

    // Assert
    expect(state).toMatchSnapshot();
  });

  it('should update todos', () => {
    // Arrange
    const initialState: fromTodos.State = {
      data: {
        entities: {
          0.07: {
            id: 0.07,
            text: 'Learn Jest',
            creationDate: new Date(),
            completed: false,
          },
        },
        ids: [0.07],
      },
      loading: false,
    };
    const updateAction = fromTodos.updateAction({
      id: 0.07,
      text: 'Learn Angular',
    });

    // Act
    const state = fromTodos.reducer(initialState, updateAction);

    // Assert
    expect(state).toMatchSnapshot();
  });

  it('should clear completed todos', () => {
    // Arrange
    const initialState: fromTodos.State = {
      data: {
        entities: {
          0.01: {
            id: 0.01,
            text: 'Learn Angular',
            creationDate: new Date(),
            completed: true,
          },
          0.02: {
            id: 0.02,
            text: 'Learn NgRx',
            creationDate: new Date(),
            completed: true,
          },
          0.03: {
            id: 0.03,
            text: 'Learn Jest',
            creationDate: new Date(),
            completed: false,
          },
        },
        ids: [0.01, 0.02, 0.03],
      },
      loading: false,
    };
    const clearCompletedAction = fromTodos.clearCompletedAction();

    // Act
    const state = fromTodos.reducer(initialState, clearCompletedAction);

    // Assert
    expect(state).toMatchSnapshot();
  });
});
