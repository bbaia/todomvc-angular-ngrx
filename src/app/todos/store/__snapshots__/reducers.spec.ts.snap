// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos reducers should add todos 1`] = `
Object {
  "data": Object {
    "entities": Object {
      "0.1": Object {
        "completed": false,
        "creationDate": 2020-01-01T00:00:00.000Z,
        "id": 0.1,
        "text": "Learn Angular",
      },
      "0.2": Object {
        "completed": false,
        "creationDate": 2020-01-01T00:00:00.000Z,
        "id": 0.2,
        "text": "Learn NgRx",
      },
    },
    "ids": Array [
      0.1,
      0.2,
    ],
  },
  "loading": false,
}
`;

exports[`todos reducers should clear completed todos 1`] = `
Object {
  "data": Object {
    "entities": Object {
      "0.03": Object {
        "completed": false,
        "creationDate": 2020-01-01T00:00:00.000Z,
        "id": 0.03,
        "text": "Learn Jest",
      },
    },
    "ids": Array [
      0.03,
    ],
  },
  "loading": false,
}
`;

exports[`todos reducers should delete todos 1`] = `
Object {
  "data": Object {
    "entities": Object {},
    "ids": Array [],
  },
  "loading": false,
}
`;

exports[`todos reducers should load todos 1`] = `
Object {
  "data": Object {
    "entities": Object {},
    "ids": Array [],
  },
  "loading": true,
}
`;

exports[`todos reducers should load todos 2`] = `
Object {
  "data": Object {
    "entities": Object {
      "0.234567891": Object {
        "completed": true,
        "creationDate": 2020-02-01T00:00:00.000Z,
        "id": 0.234567891,
        "text": "Learn Angular",
      },
      "0.345678912": Object {
        "completed": false,
        "creationDate": 2020-03-01T00:00:00.000Z,
        "id": 0.345678912,
        "text": "Learn Jest",
      },
    },
    "ids": Array [
      0.234567891,
      0.345678912,
    ],
  },
  "loading": false,
}
`;

exports[`todos reducers should mark todos as complete 1`] = `
Object {
  "data": Object {
    "entities": Object {
      "0.07": Object {
        "completed": true,
        "creationDate": 2020-01-01T00:00:00.000Z,
        "id": 0.07,
        "text": "Learn Jest",
      },
    },
    "ids": Array [
      0.07,
    ],
  },
  "loading": false,
}
`;

exports[`todos reducers should update todos 1`] = `
Object {
  "data": Object {
    "entities": Object {
      "0.07": Object {
        "completed": false,
        "creationDate": 2020-01-01T00:00:00.000Z,
        "id": 0.07,
        "text": "Learn Angular",
      },
    },
    "ids": Array [
      0.07,
    ],
  },
  "loading": false,
}
`;
